### DISCUSSION TITLE
How to Customized midi-visualizer

### DISCUSSION THREAD

**Discussion by oneandonlyonebutyou** [cifkao/html-midi-player#40] [Q&A]
@cifkao 
Hey, what a great tool .. Really appreciate it .. 
I wonder how can I develop my own <midi-visualizer>? 

I am trying to develop an app for dyslexic people and Would like to add colors  and a clean box plot 

![image](https://user-images.githubusercontent.com/17168374/147402619-912c9ea3-3bd7-408d-bdc7-954cba1bc984.png)

Could you please help me?

**Comment by cifkao** [comment]
Hi @joseph-vedadi! There are different ways to do that. I cannot give you a recipe, but I can share some tips and pointers.

You could try modifying one of the Magenta.js visualizers – the [piano roll](https://github.com/magenta/magenta-js/blob/d8a76682abb0979b985e4b80f6b68b5123b9f8d5/music/src/core/visualizer.ts#L341-L529) should be a good starting point. ([Here](https://github.com/cifkao/html-midi-player/blob/cb0291152e6d01f1091b5f0fafe8c31c1d8983d2/src/visualizer.ts#L80-L84) is where it is used in `html-midi-player`.) It uses SVG, but in your case you might be able to just use an HTML table.

Showing measure boundaries is tricky in general because you need to calculate their time in seconds (see my answer [here](https://github.com/cifkao/html-midi-player/discussions/26#discussioncomment-1088382)). But if you already know where they are (e.g. because your tempo and time signature is constant), it's easier. You could use a trick like in [this demo](https://cifkao.github.io/metrical-pe/) where I know that beats happen every 15px, so I gave the SVG element a background gradient that makes a line 1px thick every 15px, something like:
```css
midi-visualizer .piano-roll-visualizer svg {
    background: linear-gradient(90deg, #aaaaaa77 1px, transparent 1px) left/15px repeat;
}
```

Let me know if you have other questions!
